lib_eio_posix/low_level.ml

```diff
@@ -0,0 +1,200 @@
+open Eio.Std
+
+(* There are some things that should be improved here:
+
+   - Blocking FDs (e.g. stdout) wait for the FD to become ready and then do a blocking operation.
+     This might not succeed, and will block the whole domain in that case.
+     Ideally, all blocking operations should happen in a sys-thread instead.
+
+   - Various other operations, such as listing a directory, should also be done in a sys-thread
+     to avoid high latencies in the main domain. *)
+
+type ty = Read | Write
+
+module Fd = Fd
+
+let await_readable fd =
+  Fd.use_exn "await_readable" fd @@ fun fd ->
+  Sched.enter @@ fun t k ->
+  Sched.await_readable t k fd
+
+let await_writable fd =
+  Fd.use_exn "await_writable" fd @@ fun fd ->
+  Sched.enter @@ fun t k ->
+  Sched.await_writable t k fd
+
+let rec do_nonblocking ty fn fd =
+  try fn fd with
+  | Unix.Unix_error (EINTR, _, _) -> do_nonblocking ty fn fd    (* Just in case *)
+  | Unix.Unix_error((EAGAIN | EWOULDBLOCK | EINPROGRESS), _, _) ->
+    Sched.enter (fun t k ->
+        match ty with
+        | Read -> Sched.await_readable t k fd
+        | Write -> Sched.await_writable t k fd
+      );
+    do_nonblocking ty fn fd
+
+let read fd buf start len =
+  if Fd.is_blocking fd then await_readable fd;
+  Fd.use_exn "read" fd @@ fun fd ->
+  do_nonblocking Read (fun fd -> Unix.read fd buf start len) fd
+
+let write fd buf start len =
+  if Fd.is_blocking fd then await_writable fd;
+  Fd.use_exn "write" fd @@ fun fd ->
+  do_nonblocking Write (fun fd -> Unix.write fd buf start len) fd
+
```

Comment:
Did you consider starvation ?
If you have a fiber that always reads/writes and it's always available, since now they don't even trap into the scheduler they can starve everything else.
In Oslo I was build an IO greed mechanism, basically you have to force them to yield either at every IO or at some other heuristic.

Reply:
Yeah, it's one of the "Other things to work on after merging this initial support" items:

> Ensure we yield on IO to prevent starvation.

I want to add the systhread stuff first, because that will fix some of the places by itself.

Reply:
:+1: